### PR TITLE
More lazy `extract_levels` preprocessor function

### LINKED ISSUE #2114: `extract_levels` preprocessor realizes source levels
The `extract_levels` preprocessor preprocessor function realizes the source levels. These can be as big as the data for cubes with a hybrid sigma coordinate. Example with a 15GB file where this becomes problematic:
```python
from esmvalcore.dataset import Dataset
from esmvalcore.preprocessor import extract_levels
ds = Dataset(
  short_name='clw',
  mip='Amon',
  project='CMIP6',
  dataset='CESM2-WACCM',
  exp='historical',
  ensemble='r1i1p1f1',
  grid='gn',
  timerange='1850/2014',
)
cube = ds.load()
result = extract_levels(
  cube,
  levels=[0, 100000],
  scheme='linear',
  coordinate='air_pressure',
)
```

### LINKED ISSUE #2121: `extract_levels` is not lazy when using altitude - pressure levels conversion
The `extract_levels` preprocessor function is not lazy when it is used with the automatic conversion of pressure levels to altitude and vice versa.

This is because the (deprecated) `interp1d` function from scipy is not lazy: https://github.com/ESMValGroup/ESMValCore/blob/f5aef4ae6d6c5e43fc5047f652259456a62ee8eb/esmvalcore/cmor/_fixes/shared.py#L270

### PR DESCRIPTION
**Pull request by bouweandela**
## Description

Closes #2114 and closes #2121

***

## [Before you get started](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#getting-started)

- [x] [☝ Create an issue](https://github.com/ESMValGroup/ESMValCore/issues) to discuss what you are going to do

## [Checklist](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#checklist-for-pull-requests)

It is the responsibility of the author to make sure the pull request is ready to review. The icons indicate whether the item will be subject to the [🛠 Technical][1] or [🧪 Scientific][2] review.

[1]: https://docs.esmvaltool.org/en/latest/community/review.html#technical-review
[2]: https://docs.esmvaltool.org/en/latest/community/review.html#scientific-review

- [x] [🧪][2] The new functionality is [relevant and scientifically sound](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#scientific-relevance)
- [x] [🛠][1] This pull request has a [descriptive title and labels](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-title-and-label)
- [x] [🛠][1] Code is written according to the [code quality guidelines](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#code-quality)
- [x] [🧪][2] and [🛠][1] [Documentation](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#documentation) is available
- [x] [🛠][1] [Unit tests](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#tests) have been added
- [x] [🛠][1] Changes are [backward compatible](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#backward-compatibility)
- [x] [🛠][1] The [list of authors](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#list-of-authors) is up to date
- [x] [🛠][1] All [checks below this pull request](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-checks) were successful

***

To help with the number pull requests:

- 🙏 We kindly ask you to [review](https://docs.esmvaltool.org/en/latest/community/review.html#review-of-pull-requests) two other [open pull requests](https://github.com/ESMValGroup/ESMValCore/pulls) in this repository

### CODE REVIEW

**Comment by valeriupredoi** [comment]
cheers muchly for starting this @bouweandela :beer: I am testing with @ledm 's used case:
```python
import iris
from esmvalcore.preprocessor._regrid import extract_levels

fn = "/badc/cmip6/data/CMIP6/CMIP/MOHC/UKESM1-0-LL/historical/r2i1p1f2/Omon/po4/gn/v20190708/po4_Omon_UKESM1-0-LL_historical_r2i1p1f2_gn_200001-201412.nc"

cube= iris.load_cube(fn)
c2 = extract_levels(cube, scheme='nearest', levels = [0.1 ,])
```

which with `main` goes through, but uses 12.7GB of mem; with this branch I am getting this:
```
Traceback (most recent call last):
  File "/home/users/valeriu/lee_blocker.py", line 7, in <module>
    c2 = extract_levels(cube, scheme='nearest', levels = [0.1 ,])
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/users/valeriu/ESMValCore/esmvalcore/preprocessor/_regrid.py", line 1053, in extract_levels
    result = _vertical_interpolate(
             ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/users/valeriu/ESMValCore/esmvalcore/preprocessor/_regrid.py", line 875, in _vertical_interpolate
    new_data = stratify.interpolate(
               ^^^^^^^^^^^^^^^^^^^^^
  File "src/stratify/_vinterp.pyx", line 544, in stratify._vinterp.interpolate
  File "/home/users/valeriu/miniconda3-June2021/envs/esmvaltool-latest/lib/python3.11/site-packages/dask/array/core.py", line 4575, in asarray
    return from_array(a, getitem=getter_inline, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/users/valeriu/miniconda3-June2021/envs/esmvaltool-latest/lib/python3.11/site-packages/dask/array/core.py", line 3487, in from_array
    token = tokenize(x, chunks, lock, asarray, fancy, getitem, inline_array)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/users/valeriu/miniconda3-June2021/envs/esmvaltool-latest/lib/python3.11/site-packages/dask/base.py", line 999, in tokenize
    hasher = _md5(str(tuple(map(normalize_token, args))).encode())
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/users/valeriu/miniconda3-June2021/envs/esmvaltool-latest/lib/python3.11/site-packages/dask/utils.py", line 642, in __call__
    return meth(arg, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/users/valeriu/miniconda3-June2021/envs/esmvaltool-latest/lib/python3.11/site-packages/dask/base.py", line 1288, in normalize_array
    data = hash_buffer_hex(x.ravel(order="K").view("i1"))
                           ^^^^^^^^^^^^^^^^^^
numpy.core._exceptions._ArrayMemoryError: Unable to allocate 11.9 GiB for an array with shape (1603800000,) and data type float64
```

File: esmvalcore/preprocessor/_regrid.py [hunk 1]
```diff
@@ -851,28 +851,87 @@ def _create_cube(src_cube, data, src_levels, levels):
...
 
 
+def is_lazy_masked_data(array):
+    """Similar to `iris._lazy_data.is_lazy_masked_data`."""
+    return isinstance(array, da.Array) and isinstance(
+        da.utils.meta_from_array(array), np.ma.MaskedArray)
+
+
+def broadcast_to_shape(array, shape, chunks, dim_map):
+    """Copy of `iris.util.broadcast_to_shape` that allows specifying chunks."""
```
Comment (bouweandela) [inline esmvalcore/preprocessor/_regrid.py:861]: TODO: open iris issue about this

**Comment by valeriupredoi** [comment]
@bouweandela your code changes are are proving to be absolutele memory-savers, bud! I tested again with @ledm 's minimal case above, and now the run goes fine, needing order 300M of memory, as compared to a needed allocation of 12G as per file :beer: :beer: :beer: @ledm you may want to grab this one to save you tons of memory (if you are not using altitude-pressure conversions which are still realized in memory :+1:

**Comment by bouweandela** [comment]
Yes, this should work for normal use cases and possibly also with the altitude - pressure conversion. However, it does not work for derived coordinates like hybrid sigma because these end up with really large chunks (the example I saw had 14 GB chunks). I suspect that is an iris issue though, but need more time to investigate.

**Comment by ledm** [comment]
This is great - thanks! It seems to fix the minimal test case. I'll check the full run next!

**Comment by ledm** [comment]
> This is great - thanks! It seems to fix the minimal test case. I'll check the full run next!

Nope, failed again with the same error. Starting to think this is something else?

File: esmvalcore/preprocessor/_regrid.py (hunk 1 above)
Comment (bouweandela) [reply esmvalcore/preprocessor/_regrid.py:861]: Done in https://github.com/SciTools/iris/issues/5456

**Comment by bouweandela** [comment]
>  it does not work for derived coordinates like hybrid sigma because these end up with really large chunks (the example I saw had 14 GB chunks). I suspect that is an iris issue though, but need more time to investigate.

Opened an issue about this in Iris: https://github.com/SciTools/iris/issues/5457

**Comment by bouweandela** [comment]
> OK how about we merge this as is (needs a couple tests, I can do that) with the caveat emptor to be solved that derived coords need lazy-ing?

It would be great if you have time to finish this pull request

**Comment by valeriupredoi** [comment]
OK I'll have a look this week then, bud :beer:

**Comment by bouweandela** [comment]
@schlunma @valeriupredoi @ledm I think I have now fixed all the issues with making the `extract_levels` preprocessor function lazy. Could you have another go at testing this branch please and let me know if there's any further issues? Please make sure to use the [Distributed scheduler](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/quickstart/configure.html#dask-distributed-configuration) to avoid running out of memory.

**Review by schlunma — COMMENTED** [review]
Thanks @bouweandela, code looks good!

Will test this with real data now.

**Comment by schlunma** [comment]
Just tested this with a real recipe (see below) that covers all cases (regular pressure level coordinate, hybrid height coordinate, conversion altitude -> air pressure). Everything works well! 🚀 Without this PR, the recipe hangs indefinitely; with this PR, the recipe finishes in ~2 min when a reasonably high amount of memory is used:

```yaml
cluster:
  type: distributed.LocalCluster
  n_workers: 3
  threads_per_worker: 4
  memory_limit: 10 GiB
```

<details>
<summary>Recipe</summary>

```yaml
# ESMValTool
---
documentation:
  title: Test
  description: test
  authors:
    - schlund_manuel


datasets:
  - {project: CMIP6, dataset: MPI-ESM1-2-HR,   exp: historical, ensemble: r1i1p1f1, grid: gn}
  - {project: CMIP6, dataset: MPI-ESM1-2-LR,   exp: historical, ensemble: r1i1p1f1, grid: gn}
  - {project: CMIP6, dataset: HadGEM3-GC31-LL, exp: historical, ensemble: r1i1p1f3, grid: gn}
  - {project: native6, dataset: ERA5, type: reanaly, version: v1, tier: 3}


preprocessors:

  test:
    extract_levels:
      levels: 12345  # use weird level to ensure interpolation
      scheme: linear
      coordinate: air_pressure


diagnostics:

  test:
    variables:
      cl:
        timerange: '2003/2007'
        mip: Amon
        preprocessor: test
    scripts:
      null
```
</details>

File: esmvalcore/preprocessor/_regrid.py [hunk 2]
```diff
@@ -828,28 +828,90 @@ def _create_cube(src_cube, data, src_levels, levels):
     return result
 
 
+def is_lazy_masked_data(array):
+    """Similar to `iris._lazy_data.is_lazy_masked_data`."""
+    return isinstance(array, da.Array) and isinstance(
+        da.utils.meta_from_array(array), np.ma.MaskedArray)
+
+
+def broadcast_to_shape(array, shape, dim_map, chunks=None):
```
[the remark this answers is not in the document]
Comment (bouweandela) [reply esmvalcore/preprocessor/_regrid.py:837]: If it is included in the upcoming [3.8 release](https://github.com/SciTools/iris/discussions/5363) we can wait, but if it's going into the next release we are probably better off just including it here and removing it again once it has made its way into Iris.

File: esmvalcore/preprocessor/_regrid.py [hunk 3]
```diff
@@ -919,42 +981,74 @@ def parse_vertical_scheme(scheme):
     return scheme, extrap_scheme
 
 
-def extract_levels(cube,
-                   levels,
-                   scheme,
-                   coordinate=None,
-                   rtol=1e-7,
-                   atol=None):
+def _rechunk_aux_factory_dependencies(
```
[the remark this answers is not in the document]
Comment (bouweandela) [reply esmvalcore/preprocessor/_regrid.py:984]: If it is included in the upcoming [3.8 release](https://github.com/SciTools/iris/discussions/5363) we can wait, but if it's going into the next release we are probably better off just including it here and removing it again once it has made its way into Iris.

**Review by schlunma — APPROVED** [review]
Thanks for implementing my comments. I am fine with merging this now, but we could also wait until the iris features are available. Please just make sure to remove any duplicated code when the corresponding iris counterpart is available.

**Pull request merged**